### PR TITLE
add notifications on mac

### PR DESCRIPTION
**Pull request by 3nids**
kudos to @mbernasocchi for the continuous support for such a feature!
what a syntax!

### CODE REVIEW

File: src/native/mac/qgsmacnative.h [hunk 1]
```diff
@@ -19,18 +19,28 @@
 #define QGSMACNATIVE_H
 
 #include "qgsnative.h"
+#include <QPixmap>
```
Comment (m-kuhn) [inline src/native/mac/qgsmacnative.h]: Is this required?

File: src/native/mac/qgsmacnative.h [hunk 2]
```diff
@@ -19,18 +19,28 @@
...
 
 class QString;
 
 
 class NATIVE_EXPORT QgsMacNative : public QgsNative
 {
   public:
-    virtual ~QgsMacNative();
+    explicit QgsMacNative();
+    virtual ~QgsMacNative() override;
```
Comment (m-kuhn) [inline src/native/mac/qgsmacnative.h]: Remove the virtual, it's the convention

File: src/native/mac/qgsmacnative.mm [hunk 3]
```diff
@@ -18,10 +18,43 @@
...
+@interface QgsUserNotificationCenterDelegate : NSObject <NSUserNotificationCenterDelegate>
+@end
+
+@implementation QgsUserNotificationCenterDelegate
+
+- ( BOOL )userNotificationCenter:( NSUserNotificationCenter * )center shouldPresentNotification:( NSUserNotification * )notification
+{
+#pragma unused (notification)
+#pragma unused (center)
+  return YES;
```
Comment (m-kuhn) [inline src/native/mac/qgsmacnative.mm]: Wow, that's really neat stuff!

**Review by m-kuhn — APPROVED** [review]
Nice!
Just some tiny issues codewise, but once that's resolved, merge immediately.

**Comment by nyalldawson** [comment]
Awesome! Good to see OSX isn't being left behind here.

**Comment by slarosa** [comment]
Very nice, thank you!

Il Mar 21 Ago 2018, 20:14 Nyall Dawson <notifications@github.com> ha
scritto:

**Comment by nyalldawson** [comment]
@3nids/@slarosa

With this pr the only feature our OSX builds are missing is the task bar progress reporting, but my initial searches led me to believe that this isn't a standard OSX feature and would be non trivial to add.

But what *other* desktop integration features do you think qgis is missing for osx?

File: src/native/mac/qgsmacnative.h [hunk 4]
```diff
@@ -19,18 +19,28 @@
...
 
     virtual const char *currentAppLocalizedName();
     void currentAppActivateIgnoringOtherApps() override;
     void openFileExplorerAndSelectFile( const QString &path ) override;
+    QgsNative::Capabilities capabilities() const override;
+    QgsNative::NotificationResult showDesktopNotification( const QString &summary, const QString &body, const NotificationSettings &settings ) override;
+
+  private:
+    class QgsUserNotificationCenter;
+    QgsUserNotificationCenter *mQgsUserNotificationCenter;
```
Comment (nyalldawson) [inline src/native/mac/qgsmacnative.h]: = nullptr

File: src/native/mac/qgsmacnative.mm [hunk 5]
```diff
@@ -18,10 +18,43 @@
...
+}
+
+@end
+
+class QgsMacNative::QgsUserNotificationCenter
+{
+  public:
+    QgsUserNotificationCenterDelegate *_qgsUserNotificationCenter;
```
Comment (nyalldawson) [inline src/native/mac/qgsmacnative.mm]: Do you need to (can you?) initialize here? Is this a unique_ptr candidate?

File: src/native/mac/qgsmacnative.mm [hunk 6]
```diff
@@ -41,3 +74,38 @@
...
+}
+
+QgsNative::NotificationResult QgsMacNative::showDesktopNotification( const QString &summary,
+    const QString &body,
+    const QgsNative::NotificationSettings &settings )
+{
+  NSUserNotification *notification = [[NSUserNotification alloc] init];
+  notification.title = summary.toNSString();
+  notification.informativeText = body.toNSString();
+  notification.soundName = NSUserNotificationDefaultSoundName;   //Will play a default sound
```
Comment (nyalldawson) [inline src/native/mac/qgsmacnative.mm]: Is this normal behaviour on osx? Or will a sound become annoying quickly?

File: src/native/mac/qgsmacnative.mm (hunk 5 above)
Comment (3nids) [reply src/native/mac/qgsmacnative.mm]: nope, I've tried it but you cannot use unique_ptr on the objective c implementation

image

File: src/native/mac/qgsmacnative.mm (hunk 6 above)
Comment (3nids) [reply src/native/mac/qgsmacnative.mm]: yep, but then you can disable them in the notifications settings

image

**Review by 3nids — COMMENTED** [review]
@nyalldawson thanks for the review. 
For progress bar, I guess the idea would be to embed it in the application icon in the dock, I'll maybe have a look during the hackfest.
Otherwise, nothing comes to my mind in terms of os integration.

**Comment by nirvn** [comment]
> With this pr the only feature our OSX builds are missing is the task bar progress reporting, but my initial searches led me to believe that this isn't a standard OSX feature and would be non trivial to add.

@nyalldawson , @3nids , I was thinking about that overnight. While I'm not aware of OSX having a progress bar function, it _does_ have a badge count (which linux also supports). We could add a native setBadgeCount(), with an OSX and linux implentation.

On OSX, while it doesn't fully replace a progress bar, it does offer a visual indicator of an ongoing task undertaken.

**Comment by nyalldawson** [comment]
@nirvn 
> I was thinking about that overnight. While I'm not aware of OSX having a progress bar function, it does have a badge count (which linux also supports). We could add a native setBadgeCount(), with an OSX and linux implentation.

Windows 10 does to, if it's fully updated. I'd say we could do something like:

- does the platform support progress reports? if so, show them and only show badges if badge count > 1
- if not, show badge count whenever there's a task running, even if it's the only one

**Comment by nirvn** [comment]
@nyalldawson , yep, sounds like an excellent logic.

**Pull request merged**